exec/java-exec/src/main/java/org/apache/drill/exec/compile/ClassBuilder.java

```diff
@@ -0,0 +1,189 @@
+ * cg.plainOldJavaCapable(true); // Class supports plain-old Java</pre>
+ * <p>
+ * The setting to prefer plain-old Java is ignored for generated
+ * classes not marked as plain-old Java capable.
+ */
+
+public class ClassBuilder {
+
+  public static final String SAVE_CODE_OPTION = CodeCompiler.COMPILE_BASE + ".save_source";
+  public static final String CODE_DIR_OPTION = CodeCompiler.COMPILE_BASE + ".code_dir";
+
+  private final DrillConfig config;
+  private final OptionManager options;
+  private final boolean saveCode;
+  private final File codeDir;
+
+  public ClassBuilder(DrillConfig config, OptionManager optionManager) {
+    this.config = config;
+    options = optionManager;
+
+    // The option to save code is a boot-time option because
+    // it is used selectively during debugging, but can cause
+    // excessive I/O in a running server if used to save all code.
+
+    saveCode = config.getBoolean(SAVE_CODE_OPTION);
+    codeDir = new File(config.getString(CODE_DIR_OPTION));
+  }
+
+  /**
+   * Given a code generator which has already generated plain-old Java
+   * code, compile the code, create a class loader, and return the
+   * resulting Java class.
+   *
+   * @param cg a plain-old Java capable code generator that has generated
+   * plain-old Java code
+   * @return the class that the code generator defines
+   * @throws ClassTransformationException
+   */
+
+  public Class<?> getImplementationClass(CodeGenerator<?> cg) throws ClassTransformationException {
+    try {
+      return compileClass(cg);
+    } catch (CompileException | ClassNotFoundException|IOException e) {
+      throw new ClassTransformationException(e);
+    }
+  }
+
+  /**
+   * Performs the actual work of compiling the code and loading the class.
+   *
+   * @param cg the code generator that has built the class(es) to be generated.
+   * @return the class, after code generation and (if needed) compilation.
+   * @throws IOException if an error occurs when optionally writing code to disk.
+   * @throws CompileException if the generated code has compile issues.
+   * @throws ClassNotFoundException if the generated code references unknown classes.
+   * @throws ClassTransformationException generic "something is wrong" error from
+   * Drill class compilation code.
+   */
+  private Class<?> compileClass(CodeGenerator<?> cg) throws IOException, CompileException, ClassNotFoundException, ClassTransformationException {
+
+    // Get the plain-old Java code.
+
+    String code = cg.getGeneratedCode();
+
+    // Get the class names (dotted, file path, etc.)
+
+    String className = cg.getMaterializedClassName();
+    ClassTransformer.ClassNames name = new ClassTransformer.ClassNames(className);
+
+    // A key advantage of this method is that the code can be
+    // saved and debugged, if needed.
+
+    saveCode(code, name);
+
+    // Compile the code and load it into a class loader.
+
+    CachedClassLoader classLoader = new CachedClassLoader();
```

Comment:
Please see note above. Here is what we have:

```
Global class cache --> class --> class loader
```
The above pattern is the same in the byte-code fixup path and in the "plain old Java" path. The Class Loader is Java's way to get at the class.

Why we have one per class? That is the original design and it was replicated here. An advantage is that we don't actually need to add a numeric class suffix (though we do anyway) as each class loader creates its own name space.